core/vm/precompile_manager.go

```diff
@@ -0,0 +1,177 @@
+
+	// Extract the method ID from the input
+	methodId := methodID(input)
+	// Try to get the method from the precompiled contracts using the method ID
+	method, exists := pm.pMethods[addr][methodId]
+	if !exists {
+		return nil, 0, fmt.Errorf("no method with id %v in precompiled contract at address %v", methodId, addr.Hex())
+	}
+
+	gasCost := contract.RequiredGas(input)
+	if gasCost > suppliedGas {
+		return nil, 0, ErrOutOfGas
+	}
+
+	// Unpack the input arguments using the ABI method's inputs
+	unpackedArgs, err := method.abiMethod.Inputs.Unpack(input[4:])
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Convert the unpacked args to reflect values.
+	reflectedUnpackedArgs := make([]reflect.Value, 0, len(unpackedArgs))
+	for _, unpacked := range unpackedArgs {
+		reflectedUnpackedArgs = append(reflectedUnpackedArgs, reflect.ValueOf(unpacked))
+	}
+
+	ctx := precompile.NewStatefulContext(pm.evm.StateDB, addr, caller, value)
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This also makes sure that the readOnly flag isn't removed for child calls.
+	if readOnly && !ctx.IsReadOnly() {
+		ctx.SetReadOnly(true)
+		defer func() { ctx.SetReadOnly(false) }()
+	}
+
+	results := method.reflectMethod.Func.Call(append(
+		[]reflect.Value{
+			reflect.ValueOf(contract),
+			reflect.ValueOf(ctx),
+		},
+		reflectedUnpackedArgs...,
+	))
+
+	// check if precompile returned an error
+	if len(results) > 0 {
+		if err, ok := results[len(results)-1].Interface().(error); ok && err != nil {
+			return nil, 0, err
+		}
+	}
+
+	// Pack the result
+	var output []byte
+	if len(results) > 1 {
+		interfaceArgs := make([]interface{}, len(results)-1)
+		for i, v := range results[:len(results)-1] {
+			interfaceArgs[i] = v.Interface()
+		}
+		output, err = method.abiMethod.Outputs.Pack(interfaceArgs...)
+		if err != nil {
+			return nil, 0, err
+		}
+	}
+
+	suppliedGas -= gasCost
+	return output, suppliedGas, nil
+}
+
+func (pm *precompileManager) Register(addr common.Address, p precompile.StatefulPrecompiledContract) error {
+	if _, isEvmPrecompile := pm.evm.precompile(addr); isEvmPrecompile {
+		return fmt.Errorf("precompiled contract already exists at address %v", addr.Hex())
+	}
+
+	if _, exists := pm.precompiles[addr]; exists {
+		return fmt.Errorf("precompiled contract already exists at address %v", addr.Hex())
+	}
+
+	// niaeve implementation; parsed abi method names must match precompile method names 1:1
+	//
+	// Note on method naming:
+	// Method name is the abi method name used for internal representation. It's derived from
+	// the abi raw name and a suffix will be added in the case of a function overload.
+	//
+	// e.g.
+	// These are two functions that have the same name:
+	// * foo(int,int)
+	// * foo(uint,uint)
+	// The method name of the first one will be resolved as Foo while the second one
+	// will be resolved as Foo0.
+	//
+	// Alternatively could require each precompile to define the func mapping instead of doing this magic
```

Comment:
i think this is a lot better - the solidity/abigen stuff is unused apart from the abi, so we should remove it and just have the native code define the mapping.